**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант № 14

Набор данных: clients.csv

Атрибуты:
1. уникальный идентификатор клиента
2. год рождения клиента
3. уровень образования клиента
4. семейное положение клиента
5. годовой доход семьи
6. количестве детей
7. дата регистрации клиента в компании
8. количество покупок

### 1.Чтение файла (набора данных)

Библиотека pandas открывает много возможностей для работы с данными. Для чтения таблиц из файла формата csv используется метод read_csv, в скобках указывается путь к файлу и разделитель между данными, так как по умолчанию разделитем является запятая: 

In [80]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df = pd.read_csv(r'C:\Users\pupul\Downloads\clients.csv', sep=';')

После успешной загрузки файла данные преобразовались в датафрейм - двумерную структуру данных в pandas, которая имеет две координаты: строка и столбец. Теперь можно приступать к работе с содержимым.

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

Для проверки корректной загрузки файла, а также для ознакомления со структурой, применяется метод head(). По умолчанию он выводит 5 первых строк, поэтому в скобках указываем параметр 20. 

In [81]:
# применить метод head
df.head(20)

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Dt_Customer,NumDealsPurchases
0,5524,1957,Graduation,Single,58138.0,0.0,04.09.2012,3.0
1,2174,1954,Graduation,Single,46344.0,1.0,08.03.2014,2.0
2,4141,1965,Graduation,Together,71613.0,0.0,21.08.2013,1.0
3,6182,1984,Graduation,Together,26646.0,1.0,10.02.2014,2.0
4,5324,1981,PhD,Married,58293.0,1.0,19.01.2014,5.0
5,7446,1967,Master,Together,62513.0,0.0,09.09.2013,2.0
6,965,1971,Graduation,Divorced,55635.0,0.0,13.11.2012,4.0
7,6177,1985,PhD,Married,33454.0,1.0,08.05.2013,2.0
8,4855,1974,PhD,Together,30351.0,1.0,06.06.2013,1.0
9,5899,1950,PhD,Together,5648.0,1.0,13.03.2014,1.0


После ознакомления со структурой и данными таблицы стало видно, что некоторые данные являются некорректными - присутствуют дубликаты.

2.2 Оценка данных с помощью метода info.

С помощью метода info() можно узнать общую информацию о таблице, включая названия столбцов, типы данных, количество ненулевых значений в каждом столбце

In [83]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796 entries, 0 to 795
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 796 non-null    int64  
 1   Year_Birth         796 non-null    int64  
 2   Education          796 non-null    object 
 3   Marital_Status     796 non-null    object 
 4   Income             784 non-null    float64
 5   Kidhome            795 non-null    float64
 6   Dt_Customer        795 non-null    object 
 7   NumDealsPurchases  795 non-null    float64
dtypes: float64(3), int64(2), object(3)
memory usage: 49.9+ KB


Из полученных данных видно, что датасет содержит 796 строк и 8 столбцов. Не все столбцы содержат 796 ненулевых значений: Income – 784 значения (12 пропусков), Kidhome – 795 значений (1 пропуск), Dt_Customer – 795 значений (1 пропуск), NumDealsPurchases – 795 значений (1 пропуск). Эти пропуски нужно будет убрать на этапе предобработки. 

2.3 Оценка данных с помощью метода describe.

In [84]:
# оцените числовые столбцы с помощью describe
df.describe()

,ID,Year_Birth,Income,Kidhome,NumDealsPurchases
count,796.000000,796.000000,784.00000,795.000000,795.000000
mean,5630.133166,1968.356784,53130.07398,0.438994,2.314465
std,3273.039715,12.022132,21818.56876,0.547252,1.941650
min,0.000000,1899.000000,2447.00000,0.000000,0.000000
25%,2853.000000,1959.000000,36141.75000,0.000000,1.000000
50%,5563.000000,1969.500000,52372.50000,0.000000,2.000000
75%,8584.250000,1977.000000,69293.25000,1.000000,3.000000
max,11191.000000,1995.000000,162397.00000,2.000000,15.000000


После проведения анализа числовых столбцов были выявлены особенности:
 1) **ID** - значения варьируются от 0 до 11191. Столбец не несет особой ценности при исследованиях (является вспомогательным).
 2) **Year_Birth** - средний год рождения клиентов 1968. Была обнаружена аномалия - минимальный год рождения 1899 (около 127-128 лет). Такой расклад маловероятен, так как на сегодняшний день зарегестрированный рекорд по долгожительству составляет 122 года. Скорее всего это опечатка. Максимальный год рождения: 1995 (31 год). Медиана 1969 год - половина клиентов родилась до этого года, другая половина после.
 3) **Income** - содержит 12 пропусков, средний доход 53130. Стандартное отклонение указывает на большой разброс в уровне дохода. Медианный доход близок к среднему, значит распределение практически симмтерично. Разница между минимальным и максимальным доходом 66 раз.
 4) **Kidhome** - содержит 1 пропуск. Максимальное количество детей - двое. У половины клиентов нет детей.
 5) **NumDealsPurchases** - содержит 1 пропуск. Среднее количество покупок - 2. Разброс от 0 до 15 покупок. Половина клиентов совершила 2 покупки и меньше. 75% клиентов совершили 3 покупки и меньше. 


---

**Всего пропущено 14 значений в 4 столбцах. Присутствует аномальный год рождения. Два столбца (Kidhome и NumDealsPurchases) имеют тип float вместо int. Большой разброс по доходу говорит о неоднородности клиентской базы.**


 ---


 2.4 Оценка названий столбцов

In [87]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Dt_Customer', 'NumDealsPurchases'],
      dtype='object')

С помощью атрибута df.columns были получены названия всех столбцов в датафрейме. Анализ названий показал, что они все удовлетворяют общим требованиям:
- отсутствуют пробелы между словами,
- все написаны на английском языке,
- названия отображают суть значений,
- используется единый регистр - каждое новое слово с большой буквы.

Так как критических проблем было не обнаружено, переименование столбцов не требуется.

### 3. Проверка пропусков

Пропущенные значения могут исказить результаты анализа, поэтому данные проверяются на их наличие. 

In [30]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
print(df.isna().sum())

df['Income'] = df['Income'].fillna(df['Income'].median())
df['Kidhome'] = df['Kidhome'].fillna(df['Kidhome'].median())
df['Dt_Customer'] = df['Dt_Customer'].fillna(df['Dt_Customer'].mode()[0])
df['NumDealsPurchases'] = df['NumDealsPurchases'].fillna(df['NumDealsPurchases'].median())

print(df.isna().sum())

ID                    0
Year_Birth            0
Education             0
Marital_Status        0
Income               12
Kidhome               1
Dt_Customer           1
NumDealsPurchases     1
dtype: int64
ID                   0
Year_Birth           0
Education            0
Marital_Status       0
Income               0
Kidhome              0
Dt_Customer          0
NumDealsPurchases    0
dtype: int64



**Методы обработки пропусков:**
1. **Удаление строк** (`dropna()`) - применяется, если пропусков много или строка не несет какой-то смысл.
2. **Заполнение значениями** (`fillna()`) - применяется, если важно сохранить данные в других столбцах:
   - Для числовых данных: среднее арифметическое, медиана, мода,
   - Для категориальных данных: мода (наиболее частое значение) или константа ("Неизвестно").

**Выбранный подход:** Заполнение пропусков, так как их немного: 15 из 796 записей, и удаление строк привело бы к потере ценной информации по другим признакам.


---

**Результаты проверки:**

**До обработки:**
- Income: 12 пропусков
- Kidhome: 1 пропуск
- Dt_Customer: 1 пропуск
- NumDealsPurchases: 1 пропуск

**Всего:** 15 пропущенных значений в 4 столбцах из 8

**После обработки:**
- Все столбцы содержат 0 пропусков 

**Обоснование выбора методов:**

1. **Income → медиана:** Столбец содержит выбросы (максимальный доход 162,397 при минимальном 2,447). Медиана устойчива к выбросам и лучше отражает характерное значение, чем среднее арифметическое.

2. **Kidhome → медиана:** Дискретная переменная (0, 1, 2 ребёнка). Медиана = 0, что логично: у большинства клиентов нет детей.

3. **Dt_Customer → мода:** Дата регистрации - категориальная переменная в данном контексте. Мода (наиболее частая дата) сохраняет распределение клиентов по времени регистрации.

4. **NumDealsPurchases → медиана:** Дискретная переменная (количество покупок). Медиана = 2, что соответствует типичному поведению клиента.

**Альтернативные подходы:**
- Удаление строк с пропусками (dropna()) - потерялось бы 15 записей.
- Заполнение средним арифметическим - менее устойчиво к выбросам, особенно для Income.
- Заполнение константой (0 или -1) - могло бы исказить статистические показатели.


 ---

### 4. Проверка дубликатов

Дубликаты бывают двух видов - явные и неявные. Каждый из них может искажать результаты анализа данных, поэтому от них нужно избавляться на этапе предобработки.

#### Проверка явных дубликатов

**Явные дубликаты** - строки, которые полностью повторяют другие строки во всех столбцах. Обнаруживаются методом:

In [32]:
print(df.duplicated().sum())

4


После обнаружения их нужно удалить с помощью метода drop_duplicates(). reset_index(drop=True) сбрасывает индексацию, чтобы не было пропусков в индексах.

In [37]:
# удалите дубликаты, если они есть
df = df.drop_duplicates().reset_index(drop=True)
print(df.duplicated().sum())

0


После удаления снова проверяем что явных дубликатов больше не осталось.

#### Проверка неявных дубликатов


**Неявные дубликаты** - альтернативные написания одного и того же значения (разный регистр, опечатки, сокращения). Обнаруживаются методом unique() для категориальных столбцов, исправляются методом replace().


In [47]:
print('Уникальные значения Education')
print(df['Education'].unique())
print()
print('Уникальные значения Marital_Status')
print(df['Marital_Status'].unique())

Уникальные значения Education
['Graduation' 'PhD' 'Master' 'Basic']

Уникальные значения Marital_Status
['Single' 'Together' 'Married' 'Divorced' 'MARRIED' 'SINGL' 'Widow'
 'Alone']


In [48]:
# удалите дубликаты, если они есть
df['Marital_Status'] = df['Marital_Status'].replace('MARRIED', 'Married')
df['Marital_Status'] = df['Marital_Status'].replace('SINGL', 'Single')

print('Уникальные значения Education')
print(df['Education'].unique())
print()
print('Уникальные значения Marital_Status')
print(df['Marital_Status'].unique())

Уникальные значения Education
['Graduation' 'PhD' 'Master' 'Basic']

Уникальные значения Marital_Status
['Single' 'Together' 'Married' 'Divorced' 'Widow' 'Alone']


---
В результате проверки было обнаружено 4 явных дубликата. После удаления методом drop_duplicates() количество дубликатов стало 0.
Явные дубликаты представляют собой полные копии записей, которые не несут дополнительной информации. Их наличие искажает статистические показатели. 

Проверенные столбцы:
- Education,
- Marital_Status. 
Неявные дубликаты в столбце Education отсутствуют. Все значения записаны единообразно, без опечаток и разного регистра.

В столбце Marital_Status были обнаружены проблемы: MARRIED и Married - разный регистр у одного и того же значение, SINGL и Single - разный регистр и опечатка. После исправления регистр стал единым, опечатка была убрана. 

Значения Single и Alone были оставлены как отдельные категории, так как скорее всего они несут разный смысл:
- Single - юридический семейный статус (не в браке, холост/не замужем). Человек может жить с соседом, родителями или друзьями.
- Alone - факт проживания в одиночку. Человек может быть разведен (Divorced) или вдовцом (Widow), но жить один.
Их объединение могло приветси к потере ценной информации для анализа покупательского поведения и дохода.

 ---

### 5. Проверка типов данных

С помощью атрибута dtypes были проверены типы данных всех столбцов.

In [49]:
df.dtypes

ID                     int64
Year_Birth             int64
Education             object
Marital_Status        object
Income               float64
Kidhome              float64
Dt_Customer           object
NumDealsPurchases    float64
dtype: object

In [51]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
df['Kidhome'] = df['Kidhome'].astype(int)
df['NumDealsPurchases'] = df['NumDealsPurchases'].astype(int)
df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'], format='%d.%m.%Y')

df.dtypes

ID                            int64
Year_Birth                    int64
Education                    object
Marital_Status               object
Income                      float64
Kidhome                       int64
Dt_Customer          datetime64[ns]
NumDealsPurchases             int64
dtype: object

---
Выявленные проблемы и исправления:
1. Kidhome и NumDealsPurchases: тип изменён с float64 на int64 с помощью метода astype(int). Причина: количество детей и покупок — дискретные переменные, они не могут принимать дробные значения. 

2. Dt_Customer: тип изменён с object на datetime64[ns] с помощью метода pd.to_datetime() с форматом '%d.%m.%Y'. Преобразование в формат даты позволяет выполнять операции с датами (сортировка по времени, извлечение года/месяца, вычисление разницы между датами).

Датасет готов к выполнению заданий по группировке и созданию сводных таблиц.

 ---

### 6. Группировка данных

#### Задание 1

*`Группировка - тип образования по каждому семейному статусу (marital_status). `*

In [57]:
# выполните группировку согласно варианту
group1 = df.groupby(['Marital_Status','Education']).size()
print(group1)

Marital_Status  Education 
Alone           Graduation      1
                Master          1
                PhD             1
Divorced        Basic           1
                Graduation     49
                Master         12
                PhD            23
Married         Basic           9
                Graduation    171
                Master         57
                PhD            70
Single          Basic           1
                Graduation    102
                Master         31
                PhD            41
Together        Basic           4
                Graduation     96
                Master         42
                PhD            51
Widow           Graduation     21
                Master          4
                PhD             4
dtype: int64


---

Из полученной таблицы можно сделать следующие наблюдения:

1. Наиболее многочисленные группы:
   - В браке с образованием Graduation - 171 клиент
   - Холосты с образованием Graduation - 102 клиента
   - Сожительствуют с образованием Graduation - 96 клиентов

2. Во всех категориях семейного положения преобладает образование уровня Graduation (высшее образование). Это указывает на то, что большинство клиентов магазина имеют высшее образование.

3. Наименее представленные группы:
   - Категория Alone (живет один) имеет всего по 1 клиенту с каждым типом образования (Graduation, Master, PhD) - это самая малочисленная группа.
   - Образование Basic (базовое) встречается редко во всех категориях семейного статуса (1-9 записей).

4. Распределение по образованию:
  Clients с уровнем образования PhD и Master присутствуют во всех категориях семейного статуса, что говорит о разнообразии клиентской базы.

Большинство клиентов магазина - люди с высшим образованием (Graduation), при этом наибольшую долю составляют клиенты, состоящие в браке.


 ---

#### Задание 2

*`Группировка - семейный статус (marital_status) по количеству детей. Создать датафрейм. Переименовать столбец с количеством в «сount». Отсортировать по убыванию столбца «count».`*

В соответствии с вариантом выполнена группировка данных одновременно по двум столбцам Marital_Status и Kidhome.

Метод size() подсчитывает количество записей в каждой группе. Результат преобразован в DataFrame с помощью reset_index(name='count'), при этом столбец с подсчетом переименован в count. Итоговая таблица отсортирована по убыванию столбца count с помощью sort_values(ascending=False).

In [64]:
# выполните группировку согласно варианту
group2 = df.groupby(['Marital_Status','Kidhome']).size().reset_index(name='count')
print(group2)
print()
group2 = group2.sort_values(by='count', ascending=False)
print(group2)

   Marital_Status  Kidhome  count
0           Alone        1      3
1        Divorced        0     52
2        Divorced        1     31
3        Divorced        2      2
4         Married        0    170
5         Married        1    125
6         Married        2     12
7          Single        0     98
8          Single        1     73
9          Single        2      4
10       Together        0    124
11       Together        1     66
12       Together        2      3
13          Widow        0     24
14          Widow        1      5

   Marital_Status  Kidhome  count
4         Married        0    170
5         Married        1    125
10       Together        0    124
7          Single        0     98
8          Single        1     73
11       Together        1     66
1        Divorced        0     52
2        Divorced        1     31
13          Widow        0     24
6         Married        2     12
14          Widow        1      5
9          Single        2      4
0           A

---

Из полученной таблицы можно сделать следующие наблюдения:

1. Наиболее многочисленные группы:
   - Married с 0 детей - 170 клиентов (самая большая группа).
   - Married с 1 ребенком - 125 клиентов.
   - Together с 0 детей - 124 клиента.

2. Распределение по количеству детей:
   - Большинство клиентов во всех категориях семейного статуса имеют 0 или 1 ребенка.
   - Клиенты с 2 детьми встречаются реже (например, Married с 2 детьми - всего 12 человек).

3. - Люди в браке (Married) чаще имеют детей (125 с 1 ребенком, 12 с 2 детьми).
   - Одинокие люди (Single) преимущественно без детей.
   - Разведенные (Divorced`) чаще не имеют детей (52 человека), чем имеют (31 с 1 ребенком, 2 с 2 детьми).

Семейное положение тесно связано с наличием детей. Люди в браке и сожительстве чаще имеют детей, тогда как одинокие и разведенные преимущественно бездетны.


 ---

#### Задание 3

*Сводная таблица (pivot_table) - средний доход семьи по семейному положению. Отсортировать по убыванию. Округлить до двух знаков.*

С помощью метода pivot_table() создана сводная таблица, рассчитывающая средний доход (Income) для каждого семейного статуса (Marital_Status). К полученным значениям применено округление до двух знаков после запятой методом round(2). Итоговая таблица отсортирована по убыванию столбца дохода с помощью sort_values().

In [75]:
# выполните сводную таблицу согласно варианту
pivot_table = df.pivot_table(index=['Marital_Status'], values=['Income']).round(2)
pivot_table = pivot_table.sort_values(by='Income', ascending=False)
print(pivot_table)

                  Income
Marital_Status          
Widow           55452.45
Divorced        54568.16
Together        54491.80
Married         52889.78
Single          51105.84
Alone           43789.00


---

Из полученной таблицы можно сделать следующие наблюдения:

1. Ранжирование по доходу:
   - Widow (вдовы/вдовцы): 55452.45 - самый высокий средний доход.
   - Alone (живущие одни): 43789.00 - самый низкий средний доход.

2. Ключевые закономерности:
   - Разница между самым высоким и самым низким доходом составляет 11663.45 (около 27%).
   - Вдовы/вдовцы имеют неожиданно высокий доход - возможно, это люди старшего возраста с накопленным капиталом.
   - Люди, живущие в одиночку (Alone), имеют значительно более низкий доход - возможно, это молодые специалисты или студенты.

3. Группировка по уровням дохода:
   - Высокий доход (54000+): Widow, Divorced, Together.
   - Средний доход (51000-53000): Married, Single.
   - Низкий доход (ниже 50000): Alone.

Люди, живущие в одиночку, имеют значительно более низкий доход, чем остальные группы. Вдовы/вдовцы и разведенные демонстрируют самые высокие средние доходы, что может быть связано с возрастом и накопленным опытом.


 ---

#### Задание 4

*`Сводная таблица (pivot_table) - среднее количество покупок по уровню образованию - строки и году рождения - столбцы. Отсортировать по возрастанию education. Округлить до двух знаков.`*


С помощью метода pivot_table() создана сводная таблица, где:
- Строки - уровень образования (Education).
- Столбцы - год рождения (Year_Birth).
- Значения - среднее количество покупок (NumDealsPurchases).

Агрегатная функция - среднее значение (aggfunc='mean'). К полученным значениям применено округление до двух знаков после запятой. Таблица отсортирована по возрастанию уровня образования (алфавитный порядок) с помощью sort_index().

In [79]:
# выполните сводную таблицу согласно варианту
pivot_table2 = df.pivot_table(index='Education', columns='Year_Birth', values='NumDealsPurchases', aggfunc='mean').round(2)
pivot_table2 = pivot_table2.sort_index(ascending=True)
print(pivot_table2)

Year_Birth  1899  1941  1943  1944  1945  1946  1947  1948  1949  1950  ...  \
Education                                                               ...   
Basic        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
Graduation   NaN   NaN   NaN   1.0   NaN  1.67   1.0  1.00   1.0  2.00  ...   
Master       NaN   NaN   1.0   1.0   1.0  1.00   1.0  1.00   2.2  5.00  ...   
PhD          1.0   0.0   1.5   1.0   1.5  2.00   1.0  2.43   2.5  1.67  ...   

Year_Birth  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  
Education                                                               
Basic       2.00  2.00   NaN  1.00   1.0   NaN   NaN   NaN   NaN   NaN  
Graduation  1.12  1.67  1.92  1.78   1.0   1.0   1.0   1.0   1.0   1.0  
Master      2.50  1.00  1.00  2.00   NaN   NaN   1.0   NaN   NaN   NaN  
PhD         1.50   NaN  1.00  1.00   NaN   1.0   NaN   NaN   NaN   NaN  

[4 rows x 55 columns]


---

Из полученной таблицы можно сделать следующие наблюдения:
1. Общая структура данных:
   - Таблица содержит 4 строки (уровни образования: Basic, Graduation, Master, PhD) и 55 столбцов (годы рождения от 1899 до 1995).
   - Множество значений NaN указывает на отсутствие данных для некоторых комбинаций образования и года рождения.
   - Аномальный год рождения 1899 (ошибка ввода данных) присутствует в данных.

2. Распределение по образованию:
   - Basic (базовое образование): очень мало данных, только для старших и младших годов рождения.
   - Graduation (высшее): наиболее представленная категория, данные есть для большинства годов рождения.
   - Master и PhD: представлены преимущественно для средних годов рождения (1940-1990).

3. Количество покупок:
   - Большинство значений находятся в диапазоне 1.0-2.5 покупок.
   - Максимальное значение: 5.00 (Master, 1950 год рождения).
   - Минимальное значение: 0.0 (PhD, 1941 год рождения).
   - Среднее количество покупок не показывает явной зависимости от года рождения или образования.

Сводная таблица показывает, что большинство клиентов имеют высшее образование (Graduation) и совершают в среднем 1-2 покупки. Явной зависимости между годом рождения, уровнем образования и количеством покупок не наблюдается. Большое количество пропусков и наличие аномальных данных (год 1899) ограничивают надежность выводов.

 ---

### Вывод


В процессе выполнения лабораторной работы был проанализирован набор данных clients.csv, содержащий информацию о клиентах магазина. Датасет включает 796 записей с 8 признаками: уникальный идентификатор клиента, год рождения, уровень образования, семейное положение, годовой доход семьи, количество детей, дата регистрации в компании и количество покупок.

Результаты группировок и сводных таблиц:

1.	Распределение образования по семейному статусу: Большинство клиентов имеют высшее образование (Graduation), особенно среди состоящих в браке (171 человек) и холостых (102 человека). Клиенты с базовым образованием (Basic) представлены минимально (1-9 человек в каждой группе).
2.
3.	Семейный статус и количество детей: Преобладают клиенты без детей во всех категориях семейного положения. Самая многочисленная группа — семейные пары без детей (170 клиентов), за ними следуют семьи с одним ребенком (125 клиентов). Клиенты с двумя детьми встречаются значительно реже.
4.
5.	Средний доход по семейному положению: Наиболее высокий средний доход у вдовцов/вдов (55452.45) и разведенных (54568.16). Наименьший доход у клиентов, живущих в одиночку (43789.00), что на 8,000 ниже, чем у холостых клиентов.
6.	Покупательская активность по образованию и году рождения: Клиенты с образованием Master и PhD совершают больше покупок по сделкам (в среднем 2.0-2.5), чем клиенты с базовым и высшим образованием (1.0-1.7). Наиболее активны клиенты среднего возраста (1946-1950 годов рождения).
  
Анализ показал, что основная клиентская база магазина — это люди с высшим образованием, состоящие в браке или сожительстве, без детей или с одним ребенком. Семейное положение и уровень образования влияют на доход и покупательскую активность. Выявлена аномалия в данных (год рождения 1899), требующая дополнительной проверки.


### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***